astroid/constraint/constraint.py

```diff
@@ -0,0 +1,138 @@
+
+    @classmethod
+    def match(
+        cls: ConstraintT, node: NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> ConstraintT | None:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        Negate the constraint based on the value of negate.
+        """
+        if isinstance(expr, nodes.Compare) and len(expr.ops) == 1:
+            left = expr.left
+            op, right = expr.ops[0]
+            if op in {"is", "is not"} and (
+                matches(left, node)
+                and matches(right, cls.CONST_NONE)
+                or matches(left, cls.CONST_NONE)
+                and matches(right, node)
+            ):
+                negate = (op == "is" and negate) or (op == "is not" and not negate)
+                return cls(node=node, negate=negate)
+
+        return None
+
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
+        """Return True if this constraint is satisfied by the given inferred value."""
+        # Assume true if uninferable
+        if inferred is util.Uninferable:
+            return True
+
+        return self.negate ^ matches(inferred, nodes.Const(None))
+
+
+def matches(node1: nodes.NodeNG, node2: nodes.NodeNG) -> bool:
+    """Returns True if the two nodes match."""
+    if isinstance(node1, nodes.Name) and isinstance(node2, nodes.Name):
+        return node1.name == node2.name
+    if isinstance(node1, nodes.Attribute) and isinstance(node2, nodes.Attribute):
+        return node1.attrname == node2.attrname and matches(node1.expr, node2.expr)
+    if isinstance(node1, nodes.Const) and isinstance(node2, nodes.Const):
+        return node1.value == node2.value
+
+    return False
+
+
+def get_constraints(
+    expr: NameNodes, frame: nodes.LocalsDictNodeNG
+) -> dict[nodes.If, Constraint]:
+    """Returns the constraints for the given expression.
+
+    The returned dictionary maps the node where the constraint was generated to the
+    corresponding constraint.
+
+    Constraints are computed statically by analysing the code surrounding expr.
+    Currently this only supports constraints generated from if conditions.
+    """
+    current_node = expr
+    constraints: dict[nodes.If, Constraint] = {}
+    while current_node is not None and current_node is not frame:
+        parent = current_node.parent
+        if isinstance(parent, nodes.If):
+            branch, _ = parent.locate_child(current_node)
+            if branch == "body":
+                constraint = match_constraint(expr, parent.test)
+            elif branch == "orelse":
+                constraint = match_constraint(expr, parent.test, invert=True)
+            else:
+                constraint = None
+
+            if constraint:
+                constraints[parent] = constraint
+        current_node = parent
+
+    return constraints
+
+
+ALL_CONSTRAINTS = (NoneConstraint,)
+"""All supported constraint types."""
```

Comment:
This can be a set.

You could also explore using `__init_subclass__` to auto-register all Constraint subclasses with the base class. Although I'm not sure how it interacts with `abc.ABC`.

Reply:
It's not a frozenset : ``ALL_CONSTRAINT_CLASSES = frozenset((NoneConstraint,))``